libs/contrib/Data/Elem/Extra.idr

```diff
@@ -0,0 +1,52 @@
+module Data.Elem.Extra
+
+import Data.List
+import Data.List.Elem
+
+%default total
+
+||| Proof that an element is still inside a list if we append to it.
+public export
+elemAppLeft : (xs, ys : List a) -> (prf : Elem x xs) -> Elem x (xs ++ ys)
+elemAppLeft (x :: xs) ys Here = Here
+elemAppLeft (x :: xs) ys (There prf2) = There $ elemAppLeft xs ys prf2
+
+
+||| Proof that an element is still inside a list if we prepend to it.
+public export
+elemAppRight :  (ys, xs : List a) -> (prf : Elem x xs) -> Elem x (ys ++ xs)
+elemAppRight [] xs prf = prf
+elemAppRight (y :: ys) xs prf = There $ elemAppRight ys xs prf
+
+||| Proof that membership on append implies membership in left or right sublist.
+public export
+elemAppLorR : (xs, ys : List a)
+           -> (prf : Elem k (xs ++ ys))
+           -> Either (Elem k xs) (Elem k ys)
+elemAppLorR [] [] prf = absurd prf
+elemAppLorR [] _ prf = Right prf
+elemAppLorR (x :: xs) [] prf =
+  let eq   : (xs ++ [] = xs)  = appendNilRightNeutral xs
+      prf' : Elem k (x :: xs) = replace {p = \g => Elem k (x :: g)} eq prf
+  in Left prf'
+elemAppLorR (x :: xs) _ Here = Left Here
+elemAppLorR (x :: xs) ys (There prf) =
+  let iH : Either (Elem k xs) (Elem k ys) = elemAppLorR xs ys prf
+  in case iH of
+      (Left l) => Left $ There l
+      (Right r) => Right r
```

Comment:
This case also could be written simpler:
```idris
elemAppLorR (x :: xs) ys (There prf) = mapFst There $ elemAppLorR xs ys prf
```

Reply:
Again this comes down to whether or not it's better to have a terse proof, or one with (arguably) clear steps. Do we have a general consensus on this?

Reply:
I've written most of thoughts above, but for this particular example I can comment (keep in mind, that all of that is only a matter of taste):
- the name `iH` does not clarify anything; those that could would be `resultForTail` or something like this;
- the type if `iH` does not clarify anything either since it is exactly the type of result of the recursive call;
- pattern match of `Right` case does not change the type (so, there is not need to repack);
- the only thing the `Left` case is doing is an application of `There` to the what's in `Left`.

So, there are two simple operations -- recursive call for the tail and adding `There` for the `Left` case. The second one is what exactly `mapFst` created for (in case of `Either`, of course), so to all familiar with this function simple `mapFst` would communicate better than realizing that in this pattern match one case simply acts as identity and the second one maps over a particular function. The special name for a recursive call to a tail does not add anything just because there are no other calls to anything else.

And that's why I think that shorter piece of code here tells what's going on better.